### PR TITLE
implementation of sockets reporter which serialize sockets metrics da…

### PR DESCRIPTION
**Pull request by qiutongs**
…ta to report

Description
-----------

- serialize sockets metrics data to report based on user's input

Checklist:
----------

- [X I have tested my changes. No regression in existing tests.
- [X My code is Linted.

### CODE REVIEW

File: lib/defender/aws_iot_defender_reporter_socket.c [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    uint8_t hasTotal = ( tcpConnFlag & AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS_ESTABLISHED_TOTAL ) > 0;
+    uint8_t hasLocalPort = ( tcpConnFlag & AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS_ESTABLISHED_LOCAL_PORT ) > 0;
+    uint8_t hasRemoteAddr = ( tcpConnFlag & AWS_IOT_DEFENDER_METRICS_TCP_CONNECTIONS_ESTABLISHED_REMOTE_ADDR ) > 0;
+
+    uint8_t total = AwsIotDefenderInternal_GetEstablishedConnectionsCount();
+
+    _defenderMetricsConnection_t * pConnections = NULL;
+
+    /* 15 for IP + 1 for ":" + 5 for port + 1 terminator */
+    char remoteAddr[ 15 + 1 + 5 + 1 ] = "";
```
Comment (lt72) [inline lib/defender/aws_iot_defender_reporter_socket.c]: I would declare a macro at the top of the file to track the value 15 + 1 + 5 + 1

File: lib/defender/aws_iot_defender_reporter_socket.c [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    {
+        error = _AwsIotDefenderEncoder.closeContainer( pEncoderObject, &tcpConnectionMap );
+    }
+
+    return error;
+}
+
+/*-----------------------------------------------------------*/
+
+AwsIotSerializerError_t AwsIotDefenderInternal_GetLisTcpMetrics( AwsIotSerializerEncoderObject_t * pEncoderObject )
```
Comment (lt72) [inline lib/defender/aws_iot_defender_reporter_socket.c]: If this function is only called in this file then it should be called
_GetLisTcpMetrics...

what does LIs means? it is not obvious, so consider changing name...

File: lib/defender/aws_iot_defender_reporter_socket.c (hunk 1 above)
Comment (qiutongs) [reply lib/defender/aws_iot_defender_reporter_socket.c]: ok, I will do.

File: lib/defender/aws_iot_defender_reporter_socket.c (hunk 2 above)
Comment (qiutongs) [reply lib/defender/aws_iot_defender_reporter_socket.c]: This function is implemented in this file but used by "collector" to aggregate report.

It should be "Listening TCP Ports". I will change the name.

**Pull request merged**